Add documentation for cds-typer

daogrady opened the pull request.

renejeglinsky commented on tools/cds-typer.md:
````diff
@@ -0,0 +1,404 @@
+```
+
+## Using Emitted Types in Your Service
+The types emitted by the type generator are tightly integrated with the CDS API. The following section elucidates where the generated types are recognised by CDS.
````
Explain? Illustrates? Demonstrates? Let's use simpler language.
```suggestion
The types emitted by the type generator are tightly integrated with the CDS API. The following section illustrates where the generated types are recognized by CDS.
```

renejeglinsky reviewed (commented):
In general, I'd move the quickstart up and explain from there.
Should I give it a try in a branch based on yours and make a proposal?

chgeo commented:
>In general, I'd move the quickstart up and explain from there.

Yes, quick starts should always be first.

Should I give it a try in a branch based on yours and make a proposal?

Extra branch sounds cluttered and complicated.  Why not use this branch.  You only would need to sync with @daogrady when you push changes, so that you don't work in parallel.

daogrady commented on tools/cds-typer.md:
````diff
@@ -0,0 +1,404 @@
+### Command Line Interface (CLI) {#typer-cli}
+
+```sh
+npx @cap-js/cds-typer /home/mybookshop/db/schema.cds --outputDirectory /home/mybookshop
````
As of today, the default is to generate the files into the CWD. I am fine with leaving this parameter out although users will probably use it most of the time.

renejeglinsky replied:
So here the default is different from using VS Code? Did I get that right? If it's best practice to use this option, I'm fine with having it. I just had the feeling that things are much easier wrt configuration, if we just not show deviations from the defaults :)

daogrady replied:
That is correct! VSCode comes with a configuration option that is used for `--outputDirectory`. It just so happens to have its default set to `@cds-models` there and I therefore documented it that way.

daogrady commented on tools/cds-typer.md:
```diff
@@ -0,0 +1,404 @@
+
+The CLI offers several parameters which you can list using the `--help` parameter. 
+
+::: details You should then see the following output:
```
It is pretty wide, yes. I would really like to avoid having to maintain the possible parameters manually. I therefore opted for showing the output, which we can generate from `cds-typer` directly.
I do see your point that the output is currently not easy on the eye. I will try to make it look more like the [output of `cds help`](https://cap.cloud.sap/docs/tools/#cds-help).

renejeglinsky replied:
Does this command have an output that is worth showing?
```sh
npx @cap-js/cds-typer /home/mybookshop/db/schema.cds --outputDirectory /home/mybookshop
```

daogrady replied:
It does not output anything with the default configuration, no (you can pass in a `--logLevel` to produce output, but that's not the default).

daogrady replied:
![Screenshot 2023-07-19 at 11 18 35](https://github.com/cap-js/docs/assets/103028279/8f7f57f2-f605-4d0a-86de-657745622e60)
I have reworked the output of `cds-typer --help` to look more aesthetic. Would this be more agreeable? (Only screenshot for now to not interfere with your changes).

daogrady commented on tools/cds-typer.md:
```diff
@@ -0,0 +1,404 @@
+│   package.json
+│
+└───@cds-models
+│   └───<see above>
```
It refers to the layout [a bit up in the text](https://github.com/cap-js/docs/blob/e0cee5ab84de386510f6e9e018bc35f6d422042c/tools/cds-typer.md#about-the-emitted-type-files-emitted-type-files)
We don't have a way to add links in listings to point to that section more directly, do we?

renejeglinsky replied:
We could say "see previous section" that would already help. Or "see About Emmitting ..."
As this just refers to the layout I'd prefer "described in previous section". That give an idea for what someone should look in the previous section. While a "see previous section" doesn't give me enough information to decide if I want to look there now or not.

daogrady replied:
sounds good 👍

daogrady commented:
Sure, let's move quickstart to the top. 
@renejeglinsky I am perfectly fine with either a stacked PR or working on the same branch 👍

renejeglinsky commented:
Alright, I'll make my changes here and do it in one commit, in case it needs to be reverted. ;)

renejeglinsky commented:
Thanks for your patience. I did my restructuring and hope it makes sense to you. Please give it an honest review. Thanks!

renejeglinsky commented:
Thank you, I agree to all your points and make the edits accordingly

renejeglinsky commented:
I'm done. @chgeo You could give it another review, nevertheless I'll approve it now.

daogrady commented:
@renejeglinsky thank you so much for both your valuable input and your amendments! I believe it truly kicked the value up a notch. 🙂

FYI: I have added the prettified output of `cds-typer --help` to the [appropriate section](https://github.com/cap-js/docs/blob/doc/cds-typer/tools/cds-typer.md#command-line-interface-cli-typer-cli) and added a very brief explanation to the [handling of `undefined`](https://github.com/cap-js/docs/blob/doc/cds-typer/tools/cds-typer.md#handling-optional-properties) I had forgotten before.

renejeglinsky commented on node.js/typescript.md:
```diff
@@ -109,10 +109,23 @@ We invite you to contribute and help us complete the typings as appropriate. Sou
 
-There are already community contributions that fill in this gap. <br>
-See the [blog post by Simon Gaudeck](https://blogs.sap.com/2020/05/22/taking-cap-to-the-next-level-with-typescript/) for more details.
+You can find extensive documentation in a [dedicated chapter](../tools/cds-typer), together with a [quickstart guide](../tools/cds-typer#quickstart) to get everything up and running.
```
```suggestion
You can find extensive documentation in a [dedicated chapter](../tools/cds-typer), together with a [quickstart guide](../tools/cds-typer#cds-typer-vscode) to get everything up and running.
```
